### PR TITLE
[AutoParallel] Add sequence parallel for llama

### PR DESCRIPTION
**Pull request by GhostScreaming**
### PR types
Others

### PR changes
Others

### Description
Pcard-73145

Add sequence parallel for llama.

### CODE REVIEW

File: test/auto_parallel/hybrid_strategy/semi_auto_parallel_llama_model.py [hunk 1]
```diff
@@ -192,6 +192,26 @@ def forward(
             shape=target_key_value_shape
         )
 
+        if self.config.sequence_parallel:
+            query_states = dist.reshard(
```
Comment (JZ-LIANG) [inline test/auto_parallel/hybrid_strategy/semi_auto_parallel_llama_model.py]: reshard before projection while tranpose after projection

Comment (GhostScreaming) [reply test/auto_parallel/hybrid_strategy/semi_auto_parallel_llama_model.py]: Done, thx~

File: test/auto_parallel/hybrid_strategy/semi_auto_parallel_llama_model.py [hunk 2]
```diff
@@ -238,6 +250,12 @@ def forward(
         else:
             attn_output = outputs
 
+        if self.config.sequence_parallel:
+            attn_output = paddle.transpose(attn_output, [1, 0, 2])
+            attn_output = dist.reshard(
```
Comment (JZ-LIANG) [inline test/auto_parallel/hybrid_strategy/semi_auto_parallel_llama_model.py:257]: reshard should after out_projection

File: test/auto_parallel/hybrid_strategy/semi_auto_parallel_llama_model.py [hunk 3]
```diff
@@ -731,6 +772,13 @@ def forward(
 
         # if labels is None，means we need full output, instead of tensor_parallel_output
         logits = self.lm_head(hidden_states)
+        if self.config.sequence_parallel:
```
Comment (JZ-LIANG) [inline test/auto_parallel/hybrid_strategy/semi_auto_parallel_llama_model.py:773]: # reshard should before lm_head
        if self.config.sequence_parallel:
            hidden_states = dist.reshard(
                hidden_states, get_mesh(-1), [dist.Shard(1), dist.Replicate()]
            )
            # [S, B, H] -> [B, S, H]
            hidden_states = paddle.transpose(hidden_states, [1, 0, 2])

        logits = self.lm_head(hidden_states)

**Review by XieYunshen — APPROVED** [review]

**Review by JZ-LIANG — APPROVED** [review]
LGTM
